### PR TITLE
transform: schedule-memref-linalg pass

### PR DESCRIPTION
**Pull request by jorendumoulin**
This PR adds the first version of the schedule-memref-linalg pass.
This pass matches and transforms an operation on an accelerator template.

At this stage, this is quite barebones, and only functional for easy examples (elementwise operation on elementwise snax-alu accelerator/ gemm operation on gemm accelerator).

This document is an effort to explain its workings at the current stage.
[schedule-memref-linalg.pdf](https://github.com/user-attachments/files/16744943/schedule-memref-linalg.pdf)

Many improvements to this algorithm will come very shortly :)

### CODE REVIEW

File: compiler/transforms/schedule_memref_linalg.py [hunk 1]
```diff
@@ -0,0 +1,302 @@
...
+    For `dim` = 2, will return:
+        (d2) -> d2
+    """
+    return canonicalize_map(
+        map.replace_dims_and_symbols(
+            [AffineConstantExpr(0) for _ in range(dim)]
+            + [AffineDimExpr(i) for i in range(map.num_dims - dim)],
+            [],
+            map.num_dims - dim,
+            0,
```
Comment (JosseVanDelm) [inline compiler/transforms/schedule_memref_linalg.py:34]: Does this mean we should we also add todo's here for adding dynamic shapes?

File: compiler/transforms/schedule_memref_linalg.py [hunk 2]
```diff
@@ -0,0 +1,302 @@
...
+    op_type_rewrite_pattern,
+)
+from xdsl.rewriter import InsertPoint
+
+from compiler.util.canonicalize_affine import canonicalize_map
+
+
+def disable_dims(map: AffineMap, dim: int) -> AffineMap:
+    """
+    Returns an affine map with the first `dim` dimensions set to 0
```
Comment (JosseVanDelm) [inline compiler/transforms/schedule_memref_linalg.py:19]: wondering whether we should replace `first` with `leftmost`

File: compiler/transforms/schedule_memref_linalg.py [hunk 3]
```diff
@@ -0,0 +1,302 @@
...
+    """
+    bounds = bounds.copy()
+    bounds.insert(0, bounds.pop(dim - 1))
+    return bounds
+
+
+def split_dim(map: AffineMap, dim: int, template_bound: int) -> AffineMap:
+    """
+    Returns the bounds and a new affine map with the `dim` dimension split up into two
+    This translates to creating two for loops with adjuste bounds from one for loop
```
Comment (JosseVanDelm) [inline compiler/transforms/schedule_memref_linalg.py:72]:
```suggestion
    This translates to creating two for loops with adjusted bounds from one for loop
```

File: compiler/transforms/schedule_memref_linalg.py [hunk 4]
```diff
@@ -0,0 +1,302 @@
...
+def rotate_bounds(bounds: list[int | None], dim: int) -> list[int | None]:
+    """
+    Returns the bounds after rotating dims, as in rotate_dims
+    """
+    bounds = bounds.copy()
+    bounds.insert(0, bounds.pop(dim - 1))
+    return bounds
+
+
+def split_dim(map: AffineMap, dim: int, template_bound: int) -> AffineMap:
```
Comment (JosseVanDelm) [inline compiler/transforms/schedule_memref_linalg.py:69]: Your guide refers to this as tiling, but e.g. halide calls 2D loop splitting tiling, and 1D loop splitting just splitting.
Maybe change `tiling` `->` `splitting` there?

File: compiler/transforms/schedule_memref_linalg.py [hunk 5]
```diff
@@ -0,0 +1,302 @@
...
+    def match_and_rewrite(self, op: memref_stream.GenericOp, rewriter: PatternRewriter):
+        if any(isinstance(operand.type, stream.StreamType) for operand in op.operands):
+            # Already streamified
+            return
+
+        if not (
+            isinstance(op.library_call, builtin.StringAttr)
+            and op.library_call.data in ("snax_alu", "snax_gemm")
+        ):
+            raise NotImplementedError("panic!")
```
Comment (JosseVanDelm) [inline compiler/transforms/schedule_memref_linalg.py:128]:
```suggestion
            supported = ("snax_alu", "snax_gemm")
            isinstance(op.library_call, builtin.StringAttr)
            and op.library_call.data in supported
        ):
            raise NotImplementedError(f"Currently only the following library calls are supported: {" ,".join(supported)}")
```

File: compiler/transforms/schedule_memref_linalg.py [hunk 6]
```diff
@@ -0,0 +1,302 @@
...
+                    break
+
+                # else rotate the for loops
+                schedule = tuple(rotate_dims(map, schedule_dim + 1) for map in schedule)
+                schedule_bounds = rotate_bounds(schedule_bounds, schedule_dim + 1)
+
+            if not match:
+                raise RuntimeError("failed to match template and schedule")
+
+            # now, check bounds and design potential transoformation map
```
Comment (JosseVanDelm) [inline compiler/transforms/schedule_memref_linalg.py:184]:
```suggestion
            # now, check bounds and design potential transformation map
```

File: compiler/transforms/schedule_memref_linalg.py [hunk 7]
```diff
@@ -0,0 +1,302 @@
...
+                schedule_bounds = split_bounds(
+                    schedule_bounds, schedule_dim, template_bound
+                )
+
+        # if we get here we have now successfully determined a schedule for the operator
+        # this can be implemented with a memref streaming region
+        # the rest of this function just creates the wrapping memref streaming region
+        # the original generic op stays the same
+        # the indexing maps of that generic op are now pretty meaningless, but should
+        # not be looked at in following passes i think
```
Comment (JosseVanDelm) [inline compiler/transforms/schedule_memref_linalg.py:214]: Can you instead put this "wrapping" inside a separate function?

File: tests/filecheck/transforms/schedule-memref-linalg.mlir
Comment (JosseVanDelm) [inline tests/filecheck/transforms/schedule-memref-linalg.mlir:1]: Can you also add the test for the gemm (for now?) ?

**Review by JosseVanDelm — COMMENTED** [review]
Hi @jorendumoulin this is a super impressive PR!

It is so well-structured that I feel a bit bad giving suggestions/criticism :)

I'm not approving now, because I first want your opinion on these issues:
* It feels a bit icky to have separated functions working on bounds and maps, can we not generate a new class that keeps them together, which would only require one function implementation of e.g. `rotate`?
* Please add a test for `snax_gemm`
* Consider splitting the internal structure of the rewrite_pattern in two different functions, one that does the scheduling, and one that does the wrapping in the streaming region, I think that would improve reading quality, and it would also probably increase the clarity of possible error messages.

File: compiler/transforms/schedule_memref_linalg.py (hunk 1 above)
Comment (jorendumoulin) [reply compiler/transforms/schedule_memref_linalg.py:34]: yes 😢

File: compiler/transforms/schedule_memref_linalg.py (hunk 2 above)
Comment (jorendumoulin) [reply compiler/transforms/schedule_memref_linalg.py:19]: seems more clear yes

File: compiler/transforms/schedule_memref_linalg.py (hunk 4 above)
Comment (jorendumoulin) [reply compiler/transforms/schedule_memref_linalg.py:69]: yes!

File: tests/filecheck/transforms/schedule-memref-linalg.mlir
Comment (jorendumoulin) [reply tests/filecheck/transforms/schedule-memref-linalg.mlir:1]: yeppers

**Comment by jorendumoulin** [comment]
> Hi @jorendumoulin this is a super impressive PR!
> 
> It is so well-structured that I feel a bit bad giving suggestions/criticism :)
> 
> I'm not approving now, because I first want your opinion on these issues:
> 
>     * It feels a bit icky to have separated functions working on bounds and maps, can we not generate a new class that keeps them together, which would only require one function implementation of e.g. `rotate`?

Yes I agree, I will do that. However, my downstream dev branches have changed some of this functionality so to avoid having resolve all of the conflicts, I will do this after they stabilize if that's allright :)

> 
>     * Please add a test for `snax_gemm`
> 
>     * Consider splitting the internal structure of the rewrite_pattern in two different functions, one that does the scheduling, and one that does the wrapping in the streaming region, I think that would improve reading quality, and it would also probably increase the clarity of possible error messages.

I can definitely separate the scheduling and wrapping, but same comment as before, to avoid breaking my dev branches, I'd like to have this as a todo for later :)

**Review by JosseVanDelm — APPROVED** [review]

**Pull request merged**